polygeist/test/polygeist-opt/sycl/argpromotion.mlir

```diff
@@ -0,0 +1,91 @@
+!sycl_accessor_impl_device_1_ = !sycl.accessor_impl_device<[1], (!sycl_id_1_, !sycl_range_1_, !sycl_range_1_)>
+!sycl_item_base_2_ = !sycl.item_base<[2, true], (!sycl_range_2_, !sycl_id_2_, !sycl_id_2_)>
+!sycl_item_base_2_1 = !sycl.item_base<[2, false], (!sycl_range_2_, !sycl_id_2_)>
+!sycl_accessor_1_f32_w_gb = !sycl.accessor<[1, f32, write, global_buffer], (!sycl_accessor_impl_device_1_, !llvm.struct<(memref<?xf32, 1>)>)>
+!sycl_item_2_ = !sycl.item<[2, true], (!sycl_item_base_2_)>
+!sycl_item_2_1 = !sycl.item<[2, false], (!sycl_item_base_2_1)>
+!sycl_nd_item_2_ = !sycl.nd_item<[2], (!sycl_item_2_, !sycl_item_2_1, !sycl_group_2_)>
+
+gpu.module @device_func {
+  gpu.func @test1() kernel {
+    // CHECK-LABEL: gpu.func @test1() kernel
+    // CHECK:         [[RES:%.*]] = func.call @no_args() : () -> i32
+    // CHECK-NEXT:    func.call @recursive([[RES]]) : (i32) -> ()
+    // CHECK-NEXT:    {{.*}} = func.call @no_cand_arg({{.*}}) : (memref<?xi32>) -> memref<?xi32>
+    // CHECK-NEXT:    func.call @extern() : () -> ()
+    // CHECK-NEXT:    %memspacecast = memref.memory_space_cast %cast_1 : memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>> to memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>, 4>        
+    // CHECK-NEXT:    %c0 = arith.constant 0 : index
+    // CHECK-NEXT:    [[ARG0:%.*]] = "polygeist.subindex"(%memspacecast, %c0) : (memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>, 4>, index) -> memref<?xi32, 4>
+    // CHECK-NEXT:    %c1 = arith.constant 1 : index
+    // CHECK-NEXT:    [[ARG1:%.*]] = "polygeist.subindex"(%memspacecast, %c1) : (memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>, 4>, index) -> memref<?x!sycl_accessor_1_f32_w_gb, 4>
+    // CHECK-NEXT:    func.call @callee([[ARG0]], [[ARG1]], {{.*}}) : (memref<?xi32, 4>, memref<?x!sycl_accessor_1_f32_w_gb, 4>, memref<?x!sycl_nd_item_2_>) -> ()
+    // CHECK-NEXT:    gpu.return
+
+    %alloca = memref.alloca() : memref<1x!sycl_nd_item_2_>
+    %cast = memref.cast %alloca : memref<1x!sycl_nd_item_2_> to memref<?x!sycl_nd_item_2_>
+    %alloca_1 = memref.alloca() : memref<1x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>>
+    %cast_1 = memref.cast %alloca_1 : memref<1x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>> to memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>>
+    %alloca_2 = memref.alloca() : memref<1xi32>
+    %cast_2 = memref.cast %alloca_2 : memref<1xi32> to memref<?xi32>    
+    %res = func.call @no_args() : () -> i32
+    func.call @recursive(%res) : (i32) -> ()
+    func.call @no_cand_arg(%cast_2) : (memref<?xi32>) -> (memref<?xi32>)
+    func.call @extern() : () -> ()
+    %memspacecast = memref.memory_space_cast %cast_1 : memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>> to memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>, 4>    
+    func.call @callee(%memspacecast, %cast) : (memref<?x!llvm.struct<(i32, !sycl_accessor_1_f32_w_gb)>, 4>, memref<?x!sycl_nd_item_2_>) -> ()
```

Comment:
on top of ensuring the callee is in the last block, we should also ensure it is the last operation with no memory effects.
If there is a read of the original struct and callee modifies it, then this transformation is incorrect.